Module-4/Homework.playground/Contents.swift

```diff
@@ -1,2 +1,220 @@
+ 
+ **Hint**: Do not use if statements to complete this section.
+ 
+ Call the function twice, sending a different Animal each time.
+ */
+enum Animal: String {
+    case dog
+    case cat
+    case cow
+    case snake
+    case owl
+}
+
+extension Animal {
+    var sound: String {
+        switch self {
+        case .dog:
+            "woof"
+        case .cat:
+            "meow"
+        case .cow:
+            "moooo"
+        case .snake:
+            "sssssss"
+        case .owl:
+            "hoot"
+        }
+    }
+}
+
+func theSoundMadeBy(_ animal: Animal) {
+    print("A \(animal.rawValue) goes \(animal.sound).")
+}
+
+theSoundMadeBy(.dog)
+theSoundMadeBy(.snake)
+/*:
+ c) This question will have you creating multiple functions that will require you to use closures and collections. First, you will do some setup.
+ 
+ - Create an array of `Int` called nums with the values of 0 to 100.
+ 
+ - Create an array of `Int?` called `numsWithNil` with the following values:
+ `79, nil, 80, nil, 90, nil, 100, 72`
+ 
+ - Create an array of `Int` called `numsBy2` with values starting at 2 through 100, by 2.
+ 
+ - Create an array of `Int` called `numsBy4` with values starting at 2 through 100, by 4.
+ 
+ You can set the values of the arrays above using whatever method you find the easiest. In previous modules, you were introduced to ranges and sequences in Swift. Leveraging those in the Array initializer will allow you to create the requested arrays in a single line. Don’t let the last two break your stride.
+ 
+ - Create a function called `evenNumbersArray` that takes a parameter of `[Int]` (array of `Int`) and returns `[Int]`. The array of `Int` returned should contain all the even numbers in the array passed. Call the function passing the `nums` array and print the output.
+ 
+ - Create a function called `sumOfArray` that takes a parameter of `[Int?]` and returns an `Int`. The function should return the sum of the array values passed that are not `nil`. Call the function passing the `numsWithNil` array, and print out the results.
+ 
+ - Create a function called `commonElementsSet` that takes two parameters of `[Int]` and returns a `Set<Int>` (set of `Int`). The function will return a `Set<Int>` of the values in both arrays.
+ 
+ - Call the function `commonElementsSet` passing the arrays `numsBy2`, `numsBy4`, and print out the results.
+ */
+let nums = Array(0...100)
+let numsWithNil = [79, nil, 80, nil, 90, nil, 100, 72]
+let numsBy2 = Array(stride(from: 2, through: 100, by: 2))
+let numsBy4 = Array(stride(from: 2, through: 100, by: 4))
+
+func evenNumbersArray(_ array: [Int]) -> [Int] {
+    array.filter { $0 % 2 == 0 }
+}
+
+func sumOfArray(_ array: [Int?]) -> Int {
+    array.reduce(0) { runningTotal, element in
+        runningTotal + (element ?? 0)
+    }
```

Comment:
One way to also eliminate nil values from the array is to use a `compactMap`. Then, the `reduce` method can be used with a shorthand syntax.
For example:
```swift
return array.compactMap { $0 }.reduce(0, +)
```

Reply:
I appreciate the suggestion to use compactMap followed by reduce. Initially, I chose the reduce with a nil-coalescing operator to minimise the number of iterations through the array, as the compactMap + reduce combination would iterate twice.

After you comment, I decided to test the performance of each option:

```Swift
final class SummationTests: XCTestCase {

    let optionalSequence: [Int?] = {
        var arr: [Int?] = Array(0...100000)
        // insert some nil values throughout the array
        for idx in stride(from: 0, to: 10, by: (1...5).randomElement()!) {
            arr[idx] = nil
        }
        return arr
    }()

    func testTernaryReduce() {
        measure {
            let sum = optionalSequence.reduce(0) { runningTotal, element in
                runningTotal + (element ?? 0)
            }
        }
    }
    
    func testCompactMapVariant() {
        measure {
            let sum = optionalSequence.compactMap { $0 }.reduce(0, +)
        }
    }

}
```

The test results on my Mac indicated that the reduce with nil-coalescing was faster (0.012s) compared to the compactMap + reduce approach (0.022s). However, the absolute time difference (0.010s) is quite small and may not justify the loss of readability.

I would now choose the compactMap + reduce approach for its clarity and conciseness in this instance.